app/src/main/java/org/thoughtcrime/securesms/conversation/newmessage/NewMessageViewModel.kt

```diff
@@ -72,15 +77,20 @@ class NewMessageViewModel @Inject constructor(
 
             loadOnsJob = viewModelScope.launch(Dispatchers.IO) {
                 try {
-                    withTimeout(5.seconds) {
-                        SnodeAPI.getSessionID(onsNameOrPublicKey).get()
-                    }
-                    if (isActive) {
-                        _state.update { it.copy(loading = false) }
-                        onPublicKey(onsNameOrPublicKey)
-                    }
+                    // TODO move timeout to SnodeAPI#getSessionID
+                    SnodeAPI.getSessionID(onsNameOrPublicKey).asFlow()
+                        .timeout(30.seconds)
+                        .collectLatest {
+                            _state.update { it.copy(loading = false) }
+                            onPublicKey(onsNameOrPublicKey)
+                        }
                 } catch (e: Exception) {
-                    if (isActive) _state.update { it.copy(loading = false, error = GetString(e) { it.toMessage() }) }
+                    // JobCancellationException is thrown if we cancel the job
+                    // but it is internal, so this excludes other subclasses of CancellationException
+                    // than TimeoutCancellationException
+                    if (e is TimeoutCancellationException == e is CancellationException) {
```

Comment:
Why is there an `==` here?

Reply:
```
// JobCancellationException is thrown if we cancel the job
// but it is internal, so this excludes other subclasses of CancellationException
// than TimeoutCancellationException
```

Does it make sense?

Reply:
I'm sorry but I'm not sure I do.
Both `is` have to be true or both have to be false?
So `e` needs to both be a `TimeoutCancellationException` and a `CancellationException` at the same time OR be neither?

Reply:
hmmm... sorry, it's kinda esoteric. Maybe this is simpler:

```
if (e is TimeoutCancellationException || e !is CancellationException)
```

where

```
TimeoutCancellationException: CancellationException // we want to catch this
internal JobCancellationException: CancellationException // we ignore this but cant reference it.
```

IDK there's probably better ways to cancel a job that don't make your code throw. IDK why JobCancellation is `internal`.

Reply:
or... I made it simpler than that, catching each individually.

Reply:
Is there any reason to use a flow for this? 
Wouldn't this be simpler? At the end of the day the concept of `Promise` is closer to suspend function than a `Flow`. 

Then you can catch the `CancellationException` and rethrow it:

```kotlin
try {
  val result = withTimeoutOrNull(30_000) {
    SnodeAPI.getSession(onsNameOrPublicKey).await() // The `await` is an extension function on Promise
  }
  
  if (result == null) {
    // It's time out
  } else {
    // it's success
  }
} catch (e: CancellationException) {
  throw e
} catch (e: Exception) {
  // Error logic
}

```

Reply:
cool... I still used normal `timeout`, but followed this approach. done